### DISCUSSION TITLE
Invoking Services: event type in onDone action

### DISCUSSION THREAD

**Discussion by JuanPabloLoCoco** [statelyai/xstate#3176] [Q&A]
After updating the library, from version 4.8 to version 4.29, the library broke all my machines. When I call an action on the onDone event of the invoked service, the type is _done.invoke.<id>.fetching:invocation[0]_ . Before updating the library the type was _done.invoke.\<id\>_ . Is this the expected behavior?
![image](https://user-images.githubusercontent.com/23107097/160163906-37b7eb7b-5974-4fc2-8427-a873d1c08741.png)

**Comment by Andarist** [comment]
> before updating the library the type was done.invoke.<id>

This is not completely accurate - the previous pattern was `done.invoke.<idOrType>` and you have most likely been relying on the service type there. Remember that you can use the same service in multiple places of your machine - and with the old pattern, they all shared the same event type and thus leading to some unexpected behaviors for parallel states etc.

> After updating the library, from version 4.8 to version 4.29, the library broke all my machines.

Could you be more specific as to what exactly has been broken for you? Those event types should mostly be treated as implementation details.

For context - the change has been introduced in: https://github.com/statelyai/xstate/pull/2864

**Reply by JuanPabloLoCoco** [reply]
Thank you very much for the answer. I realized that i wasn't declaring an Id for the invoke service. So my machine was taking as event type the `done.invoke.${invokeConfig.src}`. I can fix my machines adding to the services an Id using the name of the service called as value.

**Reply by Andarist** [reply]
Could you describe how this change has affected you negatively?

**Reply by JuanPabloLoCoco** [reply]
I have all the events typed with Ts. For verifying the type of event, when I start running an action I ask for the type of the event. With this change, my machines start the action but return the current context because the type was different, affecting all the machine behavior.  After your answer, I realize that I have to set an Id for each service. This is quite weir for me, because in TS is not mandatory to have an Id in the services. For resolving this I wrote in each services two fields with the same string, the fields `id` and `src`. 
For the behavior of xstate, I think that should take the Id as event name (or the name of the src that is been called, in case the src is a string).

**Reply by Andarist** [reply]
The `id` is optional because technically you don't have to specify it - the machine is still capable of creating some event type for this event when you don't use any explicit ID.

I admit that there was not much guidance in the docs on how you should handle those `done.invoke.*` events when using TS and we didn't quite know how people are actually handling this in practice. It kinda never was our intention for you to include those types specifically in your `TEvent` but we also didn't provide any viable alternative to that.

We are slowly making our TypeScript story stronger though! Have you seen the typegen feature and how you can type those events with it? https://xstate.js.org/docs/guides/typescript.html#typing-promise-services
